src/server/server_family.cc

```diff
@@ -1300,25 +1300,40 @@ GenericError ServerFamily::DoSave(bool new_version, string_view basename, Transa
                         StrCat(GlobalStateName(state), " - can not save database")};
   }
 
+  auto set_last_save_error = [this](const auto& save_info) mutable {
+    last_save_info_.last_error = save_info.error;
+    last_save_info_.last_error_time = save_info.save_time;
+    last_save_info_.failed_duration_sec = save_info.duration_sec;
+  };
+
   {
     std::lock_guard lk(save_mu_);
     if (save_controller_) {
       return GenericError{make_error_code(errc::operation_in_progress),
                           "SAVING - can not save database"};
     }
+
     save_controller_ = make_unique<SaveStagesController>(detail::SaveStagesInputs{
         new_version, basename, trans, &service_, fq_threadpool_.get(), snapshot_storage_});
+
+    auto res = save_controller_->InitResourcesAndStart();
+
+    if (res) {
+      CHECK_EQ(res->error, true);
+      set_last_save_error(*res);
+      save_controller_.reset();
+      return res->error;
+    }
   }
 
-  detail::SaveInfo save_info = save_controller_->Save();
+  save_controller_->WaitAllSnapshots();
```

Comment:
Bonus point because now the function clearly suggests what's happening. @romange we can turn this function async easily. We could add a background fiber for that, which would turn `BGSAVE` from synchronous to asynchronous. This is not a priority but since it was mentioned in the issues I am bringing it up :)

Reply:
ok, this answers my first question.